Add support for lambda expressions inside mixins

skinny121 opened the pull request:
To explain the changes I will introduce some terms that relates in the changes made to the class file format in java 7 and how lambdas are implemented.

### Method Handle

Is an object that can directly execute a specified method, it holds and reference to the particular method and how to execute it, either with an invokevirtual, invokespecial, invokestatic or invokeinterface instruction. It is similar to java.lang.Mathod in some sense but it is a lot faster and is designed for execution rather than introspection. 

### Callsite

A Callsite is basically a holder for a method handle. They can either be constant or mutable(can change the method handle inside it).

### Invoke dynamic

A bytecode instruction that allows runtime linking. The first time the particular instruction is executed it calls what is called a bootstrap method which takes the name and description of the method as well as some constant arguments(arguments from the constant pool of the class) and return a Callsite. On that and any further execution of the invoke dynamic instruction the method inside the Callsite is invoked directly.

Lambda expressions use the invoke dynamic instruction to create the object representing the lambda(instance of the functional interface). It uses either java.lang.invoke.LambdaMetafactory.metafactory(...) or java.lang.invoke.LambdaMetafactory.altMetafactory(...) as its bootstrap method. The method behind invokedynamic instruction takes the local variables referenced in the lambda expression.

Javac also spits out a private method that contains all of the code that is within the lambda expression. And a Method Handle of that method is passed to the bootstrap method. 

So to support lambda expression the mixin transformer will transform invoke dynamic instructions in the same way as it transforms other invoke instructions as well as transforming the constant MethodHandle arguments it takes. To implement this I created an abstraction MemberRef that abstract over both Method Handles and bytecode invoke/field instructions.

Note any mixin class that has an invokedynamic instruction in it will now have the target class, class file version set to 52(java 8).

How do you guys test changes , to test it myself I added some tests inside of a sponge mixins and it works on both lambdas that don't refer to any shadow members and ones that do.

me4502 commented:
This looks nice. It'd be nice to have INVOKEDYNAMIC support anyway

me4502 commented on src/main/java/org/spongepowered/asm/mixin/transformer/MixinTargetContext.java:
```diff
@@ -350,6 +351,243 @@ private void transformConstantNode(MethodNode method, Iterator<AbstractInsnNode>
+     */
+    private void transformInvokeDynamicNode(MethodNode method, Iterator<AbstractInsnNode> iter, InvokeDynamicInsnNode dynInsn) {
+        // set class version to java 8
+        targetClass.version = 52;
```
INVOKEDYNAMIC was added in Java 7, should it be set to that rather than 8?

Mumfrey replied:
Should be a constant anyway, not a magic number.

gabizou commented:
Does this support method references as well?

JBYoshi commented:
@gabizou I think so. By looking at the bytecode, it appears that lambdas are converted to methods with method references. So this:
```java
public class Test {
    public void go() {
        new Thread(() -> {System.out.println("Hi");});
    }
}
```

Is compiled the same way as this:
```java
public class Test {
    public void go() {
        new Thread(Test::lambda$0);
    }

    public void lambda$0() {
        System.out.println("Hi");
    }
}
```

Except they have different line numbers.

skinny121 commented:
Yes it supports method references, method references are purely syntactic sugar for lambda expressions.

Mumfrey commented on src/main/java/org/spongepowered/asm/mixin/transformer/MixinTargetContext.java:
```diff
@@ -350,6 +351,243 @@ private void transformConstantNode(MethodNode method, Iterator<AbstractInsnNode>
+     * processed.
+     *
+     * @param handle Handle to transform
+     */
```
Missing some `@param` here.

Mumfrey commented on src/main/java/org/spongepowered/asm/mixin/transformer/MixinTargetContext.java:
```diff
@@ -350,6 +351,243 @@ private void transformConstantNode(MethodNode method, Iterator<AbstractInsnNode>
+     * instructions.
+     */
+    private abstract class MemberRef {
+        abstract boolean isField();
```
Add javadoc

Mumfrey commented on src/main/java/org/spongepowered/asm/mixin/transformer/MixinTargetContext.java:
```diff
@@ -350,6 +351,243 @@ private void transformConstantNode(MethodNode method, Iterator<AbstractInsnNode>
+
+        private final MethodInsnNode insn;
+
+        private StaticMethodRef(MethodInsnNode insn) {
```
Should be package private to avoid synthetic constructor.

Mumfrey commented on src/main/java/org/spongepowered/asm/mixin/transformer/MixinTargetContext.java:
```diff
@@ -350,6 +351,243 @@ private void transformConstantNode(MethodNode method, Iterator<AbstractInsnNode>
+
+        private final FieldInsnNode insn;
+
+        private StaticFieldRef(FieldInsnNode insn) {
```
Also package private rather than private.

Mumfrey commented on src/main/java/org/spongepowered/asm/mixin/transformer/MixinTargetContext.java:
```diff
@@ -350,6 +351,243 @@ private void transformConstantNode(MethodNode method, Iterator<AbstractInsnNode>
+
+        private Handle handle;
+
+        private HandleRef(Handle handle) {
```
And here too.

Mumfrey commented on src/main/java/org/spongepowered/asm/mixin/transformer/MixinTargetContext.java:
```diff
@@ -350,6 +351,243 @@ private void transformConstantNode(MethodNode method, Iterator<AbstractInsnNode>
+     * Reference to a field or method that also includes invocation
+     * instructions.
+     */
+    private abstract class MemberRef {
```
Please put inner class definitions at the top of the containing class.

Mumfrey commented on src/main/java/org/spongepowered/asm/mixin/transformer/MixinTargetContext.java:
```diff
@@ -413,27 +651,28 @@ private void updateStaticBinding(MethodNode method, MethodInsnNode insn) {
             if (superMethod.getOwner().isMixin()) {
-                throw new InvalidMixinException(this, "Invalid " + ASMHelper.getOpcodeName(insn) + " in " + this + " resolved " + insn.owner
-                        + " -> " + superMethod.getOwner() + " for " + insn.name + insn.desc);
+                throw new InvalidMixinException(this, "Invalid " + ASMHelper.getOpcodeName(methodRef.getOpcode()) + " in " + this + " resolved "
```
Maybe define `toString()` in `MemberRef` to return something prettier and then just use that.

Mumfrey commented on src/main/java/org/spongepowered/asm/mixin/transformer/MixinTargetContext.java:
```diff
@@ -496,6 +728,14 @@ public void transformDescriptor(TypeInsnNode typeInsn) {
+    private String transformDescriptor(Type type) {
+        if (type.getSort() == Type.METHOD) {
+            return this.transformMethodDescriptor(type.getDescriptor());
+        } else {
```
Redundant `else`.

Mumfrey commented:
This looks pretty good, if you could address the style concerns I've annotated that would be splendid. I like what you've done with the `MemberRef` thing, although they might be useful for other things so maybe make `MemberRef` a normal public class (the subclasses can be inner classes to `MemberRef` then maybe?).

Since the working branch needs some work anyway (the stuff for the INIT phase is currently a bodge until I have time to implement it properly) I have no issue with merging this and fixing any issues which crop up after the fact, though looking through the code it doesn't look like you're doing anything controversial and I like the general approach you have taken, it fits with the general style.

gabizou commented:
Now I wonder if we'd be able to apply the same sort of approach to anonymous classes (granted, anon classes are a bit of a separate beast altogether).

skinny121 commented:
I have fixed the style issues and have moved MemberRef into its own file and added toString()/equals()/hashcode() methods.

I think the version of the output class file should be the max of the mixin and the target class files as change in java 8 to allow interfaces to have static methods also causes problems. If an invocation of a static method on an interface is inside a class file version < java 8 it causes an verify error. And upgrading the class file version "shouldn't" break anything. But it does, upping the version of net.minecraft.server.MinecraftServer to java 8 causes an incorrect stack map frame verify error on a bytecode index of an FML patch. So there is something going on with FML patches and asm stack map frame generator. This also means that currently having a mixin of MinecraftServer with a lambda in it wouldn't work.

heni123321 commented on src/main/java/org/spongepowered/asm/mixin/transformer/MixinTargetContext.java:
```diff
@@ -333,20 +334,72 @@ private void transformTypeNode(MethodNode method, Iterator<AbstractInsnNode> ite
+     * @param handle Handle to transform
+     */
+    private Handle transformHandle(MethodNode method, Iterator<AbstractInsnNode> iter, Handle handle) {
+        this.targetClass.version = Constants.JAVA8_CLASSFILE_VERSION;
```
did you know you could use Opcodes.V1_8 from asm instead of making mixins own constant?

skinny121 replied:
Didn't even think to look at asm constants for it

Mumfrey commented on src/main/java/org/spongepowered/asm/mixin/transformer/MixinTargetContext.java:
```diff
@@ -667,7 +719,7 @@ public void preApply(String transformedName, ClassNode targetClass) {
 
     /**
-     * Called immediately after the mixin is applied to targetClass
+     * Called imediately after the mixin is applied to targetClass
```
This wasn't a typo to begin with :)

Mumfrey commented:
@skinny121 the target class version should really only be bumped if it needs to be, bumping to Java 8 must cause the verifier to use the new semantics for class verification which I guess isn't compatible with some of the "old" stack map frames, I think this is acceptable **provided that** turning on the class verifier with `mixin.debug.verify` detects this: if it doesn't then we need to seek alternative approaches, either by detecting and "upgrading" old classes, or at the very least inhibiting the mixin when it will cause breakage, since one of the core tenets of mixin is "do no harm" to the target class.

Basically it needs to be deterministic, if it's ASM-generated frames which are the issue (which I would be surprised by) then that may not be soluable, however if it's the case that *patched-in* code which was compiled for java 6 and then binary patched into the target without being transformed then I think we're on firmer ground, since we can always regenerate the stack map frames with ASM at that point.

> So there is something going on with FML patches and asm stack map frame generator.

Basically I'd like to see some verification of this, my suspicion is that it's not ASM's frame generator that's at fault, but more likely the java-6-style frames injected when FML is patched in.

> Now I wonder if we'd be able to apply the same sort of approach to anonymous classes (granted, anon classes are a bit of a separate beast altogether).

@gabizou the anon inner classes are, as you say, an entirely different beast. I believe I have a solution to that one though, it's not so much difficult as just a bit of a pain in the backside.

Mumfrey commented on src/main/java/org/spongepowered/asm/mixin/transformer/MemberRef.java:
```diff
@@ -0,0 +1,321 @@
+    /**
+     * A static reference to a method backed by an invoke instruction
+     */
+    public static final class StaticMethodRef extends MemberRef {
```
Sorry to be nit-picky, but could you rename the inner classes to just `Method`, `Field`, `Handle`, this will make their usage idiomatically `MemberRef.Field`, etc. which I think is cleaner.

skinny121 commented:
The incorrect stack map frames has been fixed, the bug was in MixinClassWriter.getCommonSuperClass method so when ASM was generating the stack map frames and needed to get the common superclass, it outputted the wrong type.

From you comment @Mumfrey it seems that you think that ASM doesn't regenerated all stack map frames for every method but it does because the big optimisation ASM has is only enabled if the classReader receives a MethodWriter instance from the visitMethod when executing accept method on the classReader. But as a tree is built from the classReader this optimisation cannot happen so ASM regenerates the stack map frames and does other unnecessary processing to write the methods. This is confirmed as some of the verify error happened in methods that contained no mixin code.   

One way to fix this the tree needs to somehow keep track of which methods are unchanged and then copy some code from ClassReader(and get around package-private fields the ClassReader uses for the optimisation).

bloodmc commented:
@skinny121 If you could come on IRC to speak with us that would be great =).
You can reach us on irc.esper.net in channel spongedev

Mumfrey commented:
> From you comment @Mumfrey it seems that you think that ASM doesn't regenerated all stack map frames for every method

You're right, I hadn't realised ASM regenerated the whole table if COMPUTE_FRAMES was set, although it makes sense that that's the case.

> The incorrect stack map frames has been fixed, the bug was in MixinClassWriter.getCommonSuperClass method so when ASM was generating the stack map frames and needed to get the common superclass, it outputted the wrong type.

Can you PR this fix directly to `master` please as well.
